src/php/ext/grpc/channel.c

```diff
@@ -162,6 +166,72 @@ void generate_sha1_str(char *sha1str, char *str, php_grpc_int len) {
   make_sha1_digest(sha1str, digest);
 }
 
+bool php_grpc_persistent_list_delete_unused_channel(char* target, target_bound_le_t* target_bound_status){
+  char *key = NULL;
+  zval *data;
+  int key_type;
+  PHP_GRPC_HASH_FOREACH_STR_KEY_VAL_START(&grpc_persistent_list, key, key_type, data)
+    if (key_type != HASH_KEY_IS_STRING || key == NULL) {
+      zend_throw_exception(spl_ce_InvalidArgumentException,
+                           "args keys must be strings", 1 TSRMLS_CC);
+      return FAILURE;
+    }
+    php_grpc_zend_resource *rsrc  = (php_grpc_zend_resource*) PHP_GRPC_HASH_VALPTR_TO_VAL(data)
+    if (rsrc == NULL) {
+      break;
+    }
+    channel_persistent_le_t* le = rsrc->ptr;
+    // Find the channel sharing the same target.
+    if (strcmp(le->channel->target, target) == 0) {
+      // ref_count=1 means that only the map holds the reference to the channel.
+      if(le->channel->ref_count == 1) {
+        php_grpc_delete_persistent_list_entry(le->channel->key,
+                                              strlen(le->channel->key)
+                                              TSRMLS_CC);
+        target_bound_status->current_count -= 1;
+        if (target_bound_status->current_count < target_bound_status->upper_bound) {
+          return true;
+        }
+      }
+    }
+  PHP_GRPC_HASH_FOREACH_END()
+  return false;
+}
+
+target_bound_le_t* update_and_get_target_upper_bound(char* target, int bound) {
+  php_grpc_zend_resource *rsrc;
+  target_bound_le_t* target_bound_status;
+  php_grpc_int key_len = strlen(target);
+  if (!(PHP_GRPC_PERSISTENT_LIST_FIND(&grpc_target_upper_bound_map, target,
+      key_len, rsrc))) {
+    // Target is not not persisted.
+    php_grpc_zend_resource new_rsrc;
+    target_bound_status = malloc(sizeof(target_bound_le_t));
+    if (bound == -1) {
+      // If the bound is not set, use 1 as default.s
+      bound = 1;
+    }
+    target_bound_status->upper_bound = bound;
+    // Init current_count with 1. It should be add 1 when the channel is successfully
+    // created and minus 1 when it is removed from the persistent list.
+    target_bound_status->current_count = 0;
```

Comment:
Just to confirm: `current_count` starts as 0, we increment and decrement when we add and remove from the persistent_list and it can never be -1 right?

The `upper_bound` on the other hand has -1 as the "uninitialized value"?

Reply:
I will add comment to it.
It won't be -1. It is [incremented](https://github.com/grpc/grpc/pull/15218/files/2b5f5a7c347af10c4ced1b9e3d7aa103bc51f091#diff-f942bdda09aac4a837cef783506850f0R287) when inserting one channel into the persistent map, and [decremented](https://github.com/grpc/grpc/pull/15218/files/2b5f5a7c347af10c4ced1b9e3d7aa103bc51f091#diff-f942bdda09aac4a837cef783506850f0R191) when deleting a channel from the persistent list.

`-1` used to skip setting the upper bound part. It doesn't make sense when user set the upper_bound to a value less than 0, thus only update the the upper_bound when a positive number is set. And `upper_bound` can be any negative value to indicate as "uninitialized value".